list.h

```diff
@@ -109,12 +111,13 @@ static inline int list_len(List *list)
     return list->len;
 }
 
-static inline void list_free(List *list)
+static inline void list_free(List *list, void (*free_impl)(void *))
 {
     for (Iter i = list_iter(list); !iter_end(i);) {
         ListNode *now = i.ptr;
         i.ptr = now->next;
-        free(now->elem);
+        if (free_impl)
+            free(now->elem);
```

Comment:
Should it be `free_impl(now->elem)`?

Reply:
I want to free ast list (like compound statement) by this function

Reply:
The parameter `free_impl` is not taken into evaluation yet.